Support serializing traits into specification extensions in OpenAPI

Xtansia opened the pull request:
## Description of changes

Adds a "meta-trait" `smithy.openapi#specificationExtension` that can be used to annotate other traits to indicate they should be serialized into specification extension (`x-*`) properties when converting to OpenAPI. This is supported on both shapes/"schemas" as well as operations. By default the extension will be named by the shape ID replacing `#` & `.` with `-` prefixed with `x-`, otherwise there is an `as:` member on the trait that allows renaming.

I believe this could be useful in bridging the gap where custom traits are currently unsupported in OpenAPI conversion, by allowing some opt in.

Adds a `JsonSchemaMapperV2` interface that extends the original, and adds an overload of `updateSchema` that accepts a context object that includes the model being converted.

## Example

Some operation:
```smithy
$version: "2.0"

namespace smithy.example

list StringList {
    member: String
}

@trait
@smithy.openapi#specificationExtension
structure structuredExtension {
    stringMember: String
    integerMember: Integer
    listMember: StringList
}

@trait
@smithy.openapi#specificationExtension(as: "x-metadata")
string stringExtension

@stringExtension("hello world")
structure SpecificationExtensionsOperationInput {
    name: String
    language: String
}

@structuredExtension(
    stringMember: "first field"
    integerMember: 17
    listMember: ["item1", "item2", "item3"]
)
@http(method: "PUT", uri: "/")
operation SpecificationExtensionsOperation {
    input: SpecificationExtensionsOperationInput
}

@aws.protocols#restJson1
service SpecificationExtensions {
    operations: [SpecificationExtensionsOperation]
}
```

Becomes:
```json
{
    "openapi": "3.0.2",
    "info": {
        "title": "SpecificationExtensions",
        "version": ""
    },
    "paths": {
        "/": {
            "put": {
                "operationId": "SpecificationExtensionsOperation",
                "requestBody": {
                    "content": {
                        "application/json": {
                            "schema": {
                                "$ref": "#/components/schemas/SpecificationExtensionsOperationRequestContent"
                            }
                        }
                    }
                },
                "responses": {
                    "200": {
                        "description": "SpecificationExtensionsOperation 200 response"
                    }
                },
                "x-smithy-example-structuredExtension": {
                    "stringMember": "first field",
                    "integerMember": 17,
                    "listMember": [
                        "item1",
                        "item2",
                        "item3"
                    ]
                }
            }
        }
    },
    "components": {
        "schemas": {
            "SpecificationExtensionsOperationRequestContent": {
                "type": "object",
                "properties": {
                    "name": {
                        "type": "string"
                    },
                    "language": {
                        "type": "string"
                    }
                },
               "x-metadata": "hello world"
            }
        }
    }
}
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

syall reviewed (changes requested):
Hi @Xtansia, thanks for the contribution! This PR adds a lot of new functionality to OpenAPI conversions which is powerful and exciting.

Overall feedback for the changes besides the comments:

- For new files, use the following copyright header:
  ```suggestion
  /*
   * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   * SPDX-License-Identifier: Apache-2.0
   */
  ```
- Some classes /  methods are missing updated javadoc comments.
- The tests should be expanded to cover:
  - Custom extension traits for all applicable shape types, or constrained via selector the shape types used to define custom extension traits.
  - Applying these custom extension traits to different parts of the smithy model, e.g. operations, input / output, members of structures and unions, etc.
- This feature and trait should be renamed to "Specification Extensions" rather than "Vendor Extensions" to match the official [OpenAPI specification](https://spec.openapis.org/oas/v3.1.0#specification-extensions).

The trait also needs a dedicated section in the "Converting Smithy to OpenAPI" guide at `docs/source-2.0/guides/converting-to-openapi.rst` similar to other trait definitions ([arbitrary example: `httpApiKeyAuth`](https://github.com/awslabs/smithy/blob/main/docs/source-2.0/spec/authentication-traits.rst?plain=1#L151-L207)). This would include:

- How to use it in a smithy model,
- Where to get the trait (the new `smithy-openapi-traits` package!),
- The expected OpenAPI conversion (similar to the [`@examples` trait conversion section](https://smithy.io/2.0/guides/converting-to-openapi.html#examples-trait-conversion)), and
- What subset of OpenAPI conversions are supported by this trait compared to [the OpenAPI Specification](https://spec.openapis.org/oas/v3.1.0) (behavior captured in the expanded tests).

For unsupported features, I believe the `jsonAdd` configuration setting documented in the ["Converting Smithy to OpenAPI" guide](https://smithy.io/2.0/guides/converting-to-openapi.html) should cover most cases.

Xtansia commented:
@syall Thank you for the super thorough feedback, I'll get to work addressing those 👍

Xtansia commented:
@syall I've had a first go at fixing all the changes you requested and drafting some documentation, would get your thoughts on anything that needs further improving.

syall commented on docs/source-2.0/guides/converting-to-openapi.rst:
```diff
@@ -937,6 +937,157 @@ The following is an example OpenAPI model for the above Smithy example value.
+
+Summary
+    Indicates a trait shape should be converted into an `OpenAPI specification extension`_.
+    Any custom trait that has been annotated with this trait, will be serialized into the OpenAPI specification using
```
```suggestion
    Any custom trait that has been annotated with this trait will be serialized into the OpenAPI specification using
```

syall commented on docs/source-2.0/guides/converting-to-openapi.rst:
```diff
@@ -937,6 +937,157 @@ The following is an example OpenAPI model for the above Smithy example value.
+      - Explicitly name the specification extension.
+        If set must begin with ``"x-"``, otherwise defaults to the target trait shape's ID normalized with hyphens and prepended with ``"x-"``.
+
+The following example defines a custom structure trait representing custom metadata, using the ``specificationExtension`` trait:
```
```suggestion
The following example defines a specification extension representing a custom metadata structure using the ``specificationExtension`` trait:
```

syall commented on docs/source-2.0/guides/converting-to-openapi.rst:
```diff
@@ -937,6 +937,157 @@ The following is an example OpenAPI model for the above Smithy example value.
+        output: GreetResponse
+    }
+
+This results in an ``x-meta`` property being added to the respective objects in the OpenAPI output, like below:
```
```suggestion
This results in an ``x-meta`` property being added to the respective objects in the OpenAPI output:
```

syall commented on docs/source-2.0/guides/converting-to-openapi.rst:
```diff
@@ -937,6 +937,157 @@ The following is an example OpenAPI model for the above Smithy example value.
+OpenAPI conversion traits
+-------------------------
+
+The ``software.amazon.smithy:smithy-openapi-traits`` package defines traits that can be used to augment the conversion
+of a Smithy model into an OpenAPI specification.
```
```suggestion
The ``software.amazon.smithy:smithy-openapi-traits`` package defines traits used to augment the conversion
of a Smithy model into an OpenAPI specification.
```

syall commented on docs/source-2.0/guides/converting-to-openapi.rst:
```diff
@@ -937,6 +937,157 @@ The following is an example OpenAPI model for the above Smithy example value.
+
+.. rubric:: Supported trait locations:
+
+Currently only a limited subset of trait target locations are supported in the conversion,
+all other locations including on members of the below shapes are ignored:
```
```suggestion
Only a subset of OpenAPI locations are supported in the conversion:
```

syall commented on docs/source-2.0/guides/converting-to-openapi.rst:
```diff
@@ -937,6 +937,157 @@ The following is an example OpenAPI model for the above Smithy example value.
+    * - Simple & Aggregate shapes
+      - `Schema object <https://spec.openapis.org/oas/v3.1.0#schema-object>`_
+
+Most other unsupported use cases can likely be covered by the :ref:`jsonAdd <generate-openapi-setting-jsonAdd>` feature of the ``smithy-openapi`` plugin.
```
```suggestion
Unsupported use cases can likely be covered by the :ref:`jsonAdd <generate-openapi-setting-jsonAdd>` feature of the ``smithy-openapi`` plugin.
```

syall commented on docs/source-2.0/guides/converting-to-openapi.rst:
```diff
@@ -937,6 +937,157 @@ The following is an example OpenAPI model for the above Smithy example value.
+The ``software.amazon.smithy:smithy-openapi-traits`` package defines traits that can be used to augment the conversion
+of a Smithy model into an OpenAPI specification.
+
+The following example shows how to add it to your Gradle build alongside the ``smithy-openapi`` plugin and ``smithy-build``:
```
```suggestion
The following example shows how to add it to your Gradle build alongside the ``smithy-openapi`` plugin:
```

syall commented on docs/source-2.0/guides/converting-to-openapi.rst:
```diff
@@ -937,6 +937,157 @@ The following is an example OpenAPI model for the above Smithy example value.
+      - Description
+    * - as
+      - ``string``
+      - Explicitly name the specification extension.
+        If set must begin with ``"x-"``, otherwise defaults to the target trait shape's ID normalized with hyphens and prepended with ``"x-"``.
```
```suggestion
      - Explicitly name the specification extension.
        If set, it must begin with ``"x-"``. Otherwise, it defaults to the target trait's shape ID normalized with hyphens and prepended with ``"x-"``.
```

syall reviewed (changes requested):
Hi @Xtansia, all of my comments are on the documentation section added.

One item I missed from last review is that `SpecificationExtensionsMapper` is missing a corresponding `SpecificationExtensionsMapperTest`. Besides this, the code changes look good!

Xtansia commented:
> Hi @Xtansia, all of my comments are on the documentation section added.
> 
> One item I missed from last review is that `SpecificationExtensionsMapper` is missing a corresponding `SpecificationExtensionsMapperTest`. Besides this, the code changes look good!

@syall Is there something specific you'd want to be covered by a `SpecificationExtensionsMapperTest` or just moving the current test I'd added out of `OpenApiConverterTest`?

Xtansia commented:
@syall I've refactored those tests and added cases covering all trait shapes on services, operation, structure and string (as an "inlined" schema). Do you think that's sufficient or would you like coverage of applying the extensions to all simple & aggregate types?

syall commented:
@Xtansia The current tests are sufficient, I believe that the Simple & Aggregate shapes fall under the same testing category for "inlined" schema.

Xtansia commented:
@syall Other possibility I can think of for not breaking current implementations of `JsonSchemaMapper` is:

Add overload with default implementation to interface:
```java
public interface JsonSchemaMapper {
    default byte getOrder() { return 0; }

    default Schema.Builder updateSchema(Shape shape, Schema.Builder schemaBuilder, JsonSchemaConfig config) { return schemaBuilder; }

    default Schema.Builder updateSchema(Shape shape, Schema.Builder schemaBuilder, JsonSchemaConfig config, Model model) { return schemaBuilder; }
}
```

And change `JsonSchemaShapeVisitor::buildSchema` to:
```java
    private Schema buildSchema(Shape shape, Schema.Builder builder) {
        JsonSchemaConfig config = converter.getConfig();
        
        for (JsonSchemaMapper mapper : mappers) {
            builder = mapper.updateSchema(shape, builder, config);
            builder = mapper.updateSchema(shape, builder, config, model);
        }

        return builder.build();
    }
```

Though it's a tad gross in duplication, and I'm not sure if adding a default implementation to the original method def is class compatible with code compiled with the previous interface version. Alternatively could do similar thing but with a new interface (ie. `JsonSchemaMapperWithContext`) that extends old `JsonSchemaMapper` and then do `instanceof` differentiation in `buildSchema`

Xtansia commented:
@syall Have pushed up an a version of what I suggested, also changed to passing in a "context" object to the new interface to minimise breaking changes in future.

Xtansia commented:
@syall Have you been able to review my alternative approach?

syall commented on smithy-jsonschema/src/test/java/software/amazon/smithy/jsonschema/JsonSchemaConverterTest.java:
Changed `JsonSchemaMapper` tests to class implementation rather than functional interface.

syall commented on smithy-jsonschema/src/main/java/software/amazon/smithy/jsonschema/JsonSchemaMapper.java:
There is a breaking change in `JsonSchemaMapper` from a [functional interface](https://www.geeksforgeeks.org/functional-interfaces-java/) to a normal interface.

Although this is a breaking change, I think this is a reasonable compromise to enable this feature.

syall reviewed (commented):
Changed the previous approach to use an overloaded `JsonSchemaMapper::updateSchema()` that uses the new `JsonSchemaMapperContext`.

`JsonSchemaMapperContext` is a class that contains all the info needed to update schemas, and is additive by adding fields to the class (e.g. in this PR `Model`). This avoids adding overloaded methods with parameters.
